samples/multiplatform/kmp-lib-sample/src/iosMain/kotlin/com/apollographql/apollo/kmpsample/data/ApolloCoroutinesService.kt

```diff
@@ -0,0 +1,65 @@
+package com.apollographql.apollo.kmpsample.data
+
+import com.apollographql.apollo.api.Operation
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.kmpsample.GithubRepositoriesQuery
+import com.apollographql.apollo.kmpsample.GithubRepositoryCommitsQuery
+import com.apollographql.apollo.kmpsample.GithubRepositoryDetailQuery
+import okio.Buffer
+import okio.toByteString
+import platform.Foundation.NSMutableURLRequest
+import platform.Foundation.NSString
+import platform.Foundation.NSURL
+import platform.Foundation.NSURLConnection
+import platform.Foundation.NSUTF8StringEncoding
+import platform.Foundation.create
+import platform.Foundation.dataUsingEncoding
+import platform.Foundation.sendSynchronousRequest
+import platform.Foundation.setHTTPBody
+import platform.Foundation.setHTTPMethod
+import platform.Foundation.setValue
+import kotlin.coroutines.suspendCoroutine
+
+actual class ApolloCoroutinesService {
+
+  private val url = NSURL(string = BASE_URL)
+
+  actual suspend fun fetchRepositories(query: GithubRepositoriesQuery): Response<GithubRepositoriesQuery.Data> =
+      suspendCoroutine {
+        doRequest(query)
+      }
+
+  actual suspend fun fetchRepositoryDetail(query: GithubRepositoryDetailQuery): Response<GithubRepositoryDetailQuery.Data> =
+      suspendCoroutine {
+        doRequest(query)
+      }
+
+  actual suspend fun fetchCommits(query: GithubRepositoryCommitsQuery): Response<GithubRepositoryCommitsQuery.Data> =
+      suspendCoroutine {
+        doRequest(query)
+      }
+
+  private fun <D : Operation.Data, T, V : Operation.Variables> doRequest(query: Operation<D, T, V>): Response<T> {
+    val request = createRequest(query)
+    val response = NSURLConnection.sendSynchronousRequest(request, null, null)
+    return query.parse(Buffer().write(response!!.toByteString()))
```

Comment:
as far as I remember KN still doesn't support multiple threaded coroutines. Means you can't offload network execution to background thread using coroutines. Coroutines work only on main thread.

But what we can do is similar to Ktor, schedule network execution on background thread parse response  and then switch back to the main thread (at least that how we using it at Shopify).

If you want I can extract network executor that we have and add to this PR.

Reply:
That would definitely be welcome.  This was me trying to learn and implement at the same time 😊